### PR TITLE
add Nextcloud 20 unified searching

### PR DESCRIPTION
**Pull request by dassio**
nextcloud change the search mechanism , using an unified search . app has to provide an class that implements `IProvider` interface

### CODE REVIEW

**Review by matiasdelellis — CHANGES_REQUESTED** [review]
Hi @dassio 
Thank you very much for the contribution.

Most of the comments are due bad styles.. but first I would recommend that you try to divide the PR in two.

One PR to support NC20.
* Please make sure it works. You tested?
* You still have to enable the version!. https://github.com/matiasdelellis/facerecognition/blob/master/appinfo/info.xml#L40
* Make sure they pass the tests. Add NC 20 https://github.com/matiasdelellis/facerecognition/blob/master/.travis.yml#L23-L25

Another PR for unified search.
* I think it is a good feature of NC20, but I think it is not intended to return all photos of a person.
* It should return the persons that you found, and if you click them, show the photos.
* For this feature, you may want to wait that i finish the person view on PR #336

File: lib/Search/PersonSearchProvider.php
Comment (dassio) [reply lib/Search/PersonSearchProvider.php:143]: it is used here : ` $this->formatSubline($result), `

File: lib/Search/PersonSearchProvider.php
Comment (dassio) [reply lib/Search/PersonSearchProvider.php:102]: no , '0' mean the highest 
I think this is unncessay , i copied this one from the files app, their roginal meaning should be  :  
if we are at the files view , the files search provider should have the  highest priority
i will delete this one

File: lib/Search/PersonSearchProvider.php
Comment (dassio) [reply lib/Search/PersonSearchProvider.php:64]: yes, $modelId is better

File: lib/AppInfo/Application.php
Comment (dassio) [reply lib/AppInfo/Application.php:45]: Ok

**Comment by dassio** [comment]
> It should return the persons that you found, and if you click them, show the photos.

not really,  you can see all the persons at the personal setting page,  the unified search is to display the photos belonging to that person 
unified search faces

i don't see a way to dispaly certain photos  after one click: nextcloud does not have this  feature to filter photos . and it wil really difficult to change the behaviour of the unified search client js (probably server guys will not accept any PR specific to one app)

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [reply lib/Search/PersonSearchProvider.php]: The unified search is shown in any application. So you should add high priority when you are in facerecognition, and low in any other. 😉

**Comment by matiasdelellis** [comment]
Hi @dassio 
There are comments that I received in the mail, but not here.

> not really, you can see all the persons at the personal setting page, the unified search is to display the photos belonging to that person

No. The unified search, which allows is to unify the search framework among all the applications!. 😅 
When you search within Files, Photos, Deck, Notes, or even within the Settings, all providers are consulted. So these queries should be fast, and since the results of all applications are displayed, they should show only the relevant information.

Can you imagine that happend if you rename your files to mom_***. Jpg, and files return 5000 files and facerecognition shows you another 5000?. It is completely insane. 🙈 😅 

That said, I insist on only showing a link to the person within our application. 😬 
This view doesn't exist yet (Depend on #336), but just opening our application is more than enough.

> i don't see a way to dispaly certain photos after one click: nextcloud does not have this feature to filter photos . and it wil really difficult to change the behaviour of the unified search client js (probably server guys will not accept any PR specific to one app)

Example..
https://github.com/nextcloud/deck/commit/2059e55e30957dfcbebe5dfa3e89a5dc7ea5bf17

**Comment by dassio** [comment]
Hi @matiasdelellis  
 sorry for the confusion, before  i did not check the PR you mentioned , then i saw it, that's why i delete the comment 

yes, it would be better to have a link to the personal setting page, i will update on that

File: lib/Db/ImageMapper.php
Comment (matiasdelellis) [inline lib/Db/ImageMapper.php:205]: Don't use it, and remove it..
I understand what you wanted to do, but this is wrong. The avatar is not a photo of a person. If you can't get the face thumnail better show the application icon. :wink:

File: lib/Db/PersonMapper.php
Comment (matiasdelellis) [inline lib/Db/PersonMapper.php:426]: Please, return the array of persons, not an array of string..

As I said, you should wait for the merge of #336  and reuse https://github.com/matiasdelellis/facerecognition/blob/bb4f7f24854d6c2b52fe55156a348e5943dc3908/lib/Db/PersonMapper.php#L132-L147

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [reply lib/Search/PersonSearchProvider.php:69]: /** @var string */

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [inline lib/Search/PersonSearchProvider.php:86]: Remember spaces.. :wink:

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [inline lib/Search/PersonSearchProvider.php:107]: Restore optional priority.
It should be a high priority if you are within the application, and low anywhere else. :smile:

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [inline lib/Search/PersonSearchProvider.php:120]: Use https://github.com/matiasdelellis/facerecognition/pull/336/commits/bb4f7f24854d6c2b52fe55156a348e5943dc3908

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [inline lib/Search/PersonSearchProvider.php:128]: As I said above, I prefer that you show the application icon. :disappointed:

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [inline lib/Search/PersonSearchProvider.php:135]: MM.. I still don't check the api to know what the parameter is, but you never return a folder.. :confused:

**Review by matiasdelellis — CHANGES_REQUESTED** [review]
Wow.. Much better .. Thanks again. :smile: 

Tomorrow I'll merge #336 and you should rebase on it to reuse a lot of code. 

Sorry if I'm annoying.. :see_no_evil: :sweat_smile:

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [reply lib/Search/PersonSearchProvider.php:128]: p.s: In another update we can add it, but now it will be more than enough, and now in line with many official applications. :smile:

File: lib/Search/PersonSearchProvider.php
Comment (dassio) [reply lib/Search/PersonSearchProvider.php:69]: corrected

File: lib/Search/PersonSearchProvider.php
Comment (dassio) [reply lib/Search/PersonSearchProvider.php:128]: i added an avatar placeholder

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [inline lib/Search/PersonSearchProvider.php:124]:
```suggestion
				    ''
```

Don't add a new icon, just use the application icon at least in this stage..

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [inline lib/Search/PersonSearchProvider.php:128]:
```suggestion
				    'icon-facerecognition',
```
Acording:
https://github.com/nextcloud/server/blob/3ffd09d3f191def63ad2766368aaef62b333fcd1/lib/public/Search/SearchResultEntry.php#L86-L107

This parameter is the class of the icon... :wink: 

MM.. You must include the icon as:
https://github.com/matiasdelellis/facerecognition/blob/b26f39a1b47f7bfe52327548e61c87dba58c9ba7/css/facerecognition.scss#L5-L7

File: lib/Service/UrlService.php
Comment (matiasdelellis) [inline lib/Service/UrlService.php:123]: Good catch.. :grimacing:

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [reply lib/Search/PersonSearchProvider.php:124]: p.s: The main reason is that bitmap icons don't adapt on dark themes.  :sweat_smile:

**Review by matiasdelellis — CHANGES_REQUESTED** [review]
Hi @dassio 

Great!!!. :smile: :smiley: :grimacing: :grin:  :grinning: :tada: 

Last change that I will ask you, and after testing we can merge it. :grinning:

File: lib/Search/PersonSearchProvider.php
Comment (matiasdelellis) [reply lib/Search/PersonSearchProvider.php:128]:
```
 .icon-facrecognition { 
 	@include icon-color('app-dark', 'facerecognition', $color-black); 
 } 
```

**Comment by matiasdelellis** [comment]
Ok.. I'm already testing NC19, and it still seems to work fine .. :smile:

**Comment by matiasdelellis** [comment]
Hi @dassio 
Please, see https://github.com/matiasdelellis/facerecognition/issues/343#issuecomment-713028656 and the comments that follow

**Comment by benjaminaigner** [comment]
@matiasdelellis @dassio 

Dear all, I'm really glad you worked out the unified search!
Is there progress in merging or do you need help in any way for this PR?
Or do you recommend just to check out this PR, apply your change requests and run it?

**Comment by matiasdelellis** [comment]
Hi @dassio 

I take your work and I'm about to publish it. :grimacing: 

https://services.delellis.com.ar/data/facerecognition/facerecognition.tar.gz

Please, test. just replace them in the apps/facerecognition folder with this.

Thanks for all again. :smile:

**Comment by matiasdelellis** [comment]
Merged in last release.. Thanks again.. :smile:

**Pull request closed without merging**
